tensorflow/compiler/mlir/hlo/include/mlir-hlo/Dialect/mhlo/transforms/map_lmhlo_to_scalar_op.h

```diff
@@ -165,15 +246,64 @@ struct isComplexType {
   bool operator()(Type t) { return t.isa<ComplexType>(); }
 };
 
+// helper type
+template <typename... T>
+struct TypeSequence {
+  constexpr static size_t size() { return sizeof...(T); }
+};
+
+// if MapTy<LhloOpTy> is legal type FilterSeqType::type same as
+// TypeSequence<SupportedType, MapTy<LhloOpTy>> 
+// otherwise FilterSeqType::type same as TypeSequence<>
+template <typename SupportedType, typename LhloOpTy,
+          template <typename T> typename MapTy,
+          typename P = llvm::is_detected<MapTy, LhloOpTy>>
+struct FilterSeqType {
+  using type = TypeSequence<>;
+};
+template <typename SupportedType, typename LhloOpTy,
+          template <typename T> typename MapTy>
+struct FilterSeqType<SupportedType, LhloOpTy, MapTy, std::true_type> {
+  using type = TypeSequence<SupportedType, MapTy<LhloOpTy>>;
+};
+
+// helper function reduction return type
+template <typename... LTys, typename... RTys>
+auto constexpr static sequence_cat(TypeSequence<LTys...>,
+                                   TypeSequence<RTys...>) {
+  return TypeSequence<LTys..., RTys...>{};
+}
+// SequenceCat<TypeSequence<T0, T1...>, TypeSequence<M0, M1...>>::type
+// is same with TypeSequence<T0, T1..., M0, M1...>
+template <typename FSeqTy, typename SSeqTy, typename... RSeqTy>
+struct SequenceCat {
+  using type = typename SequenceCat<decltype(sequence_cat(FSeqTy(), SSeqTy())),
+                                    RSeqTy...>::type;
+};
+template <typename FSeqTy, typename SSeqTy>
+struct SequenceCat<FSeqTy, SSeqTy> {
+  using type = decltype(sequence_cat(FSeqTy(), SSeqTy()));
+};
+
+template <template <typename... T> typename FuncTy, typename... ETys>
+auto InvokeTemplateArgvs(const TypeSequence<ETys...>&) {
+  return FuncTy<ETys...>{};
+}
+
 // Inserts the computation that corresponds to the body of the loop for lowered
 // LHLO unary/binary op. Returns the value for the result.
 template <typename LhloOpTy>
 inline Value MapLhloOpToStdScalarOp(Location loc, ArrayRef<Type> result_types,
                                     ArrayRef<Type> arg_types,
                                     ArrayRef<Value> args, OpBuilder* b) {
-  return MapLhloOpToScalarOpImpl<isSignedIntegerType, ScalarIOp<LhloOpTy>,
-                                 isUnsignedIntegerType, ScalarUOp<LhloOpTy>,
-                                 isFloatType, ScalarFOp<LhloOpTy>>{}(
+  // TODO(lipracer): we need check AllTemplateArguments at compile time
```

Comment:
Have you considered something simpler than this? The template meta-programming here is a bit hard to follow at first glance. For example, what if we did something like(might not be 100% exact):

```c++
struct LhloToScalarOpBase {
  using FOp = void;
  using IOp = void;
  using UOp = void;
  using COp = void; 
};

template <typename LhloBinaryOpTy>
struct LhloToScalarOp : public LhloToScalarOpBase {};
template <>
struct LhloToScalarOp<lmhlo::AddOp> : public LhloToScalarOpBase {
  using FOp = ::mlir::AddFOp;
  using IOp = ::mlir::AddIOp;
  using UOp = ::mlir::AddIOp;
  using COp = ::mlir::complex::AddOp;
};
template <>
struct LhloToScalarOp<lmhlo::AndOp> {
  using IOp = ::mlir::AndOp;
  using UOp = ::mlir::AndOp;
};
...

// Note the `void` middle param, which handles the case where the Op didn't
// provide the corresponding FOp/IOp/etc.
template <typename SupportedType, typename... Args>
struct MapLhloOpToScalarOpImpl<SupportedType, void, Args...> {
  Value operator()(Location loc, ArrayRef<Type> result_types,
                   ArrayRef<Type> arg_types, ArrayRef<Value> args,
                   OpBuilder* b) {
    return MapLhloOpToScalarOpImpl<Args...>{}(loc, result_types, arg_types,
                                              args, b);
  }
};

// Inserts the computation that corresponds to the body of the loop for lowered
// LHLO unary/binary op. Returns the value for the result.
template <typename LhloOpTy>
inline Value MapLhloOpToStdScalarOp(Location loc, ArrayRef<Type> result_types,
                                    ArrayRef<Type> arg_types,
                                    ArrayRef<Value> args, OpBuilder* b) {
  return MapLhloOpToScalarOpImpl<
    isSignedIntegerType, ScalarIOp<LhloOpTy>, 
    isUnsignedIntegerType, ScalarUOp<LhloOpTy>,
    isFloatType, ScalarFOp<LhloOpTy>,
    isComplexType, ScalarCOp<LhloOpTy>>(
      loc, result_types, arg_types, args, b);
}
```

Reply:
Done.